### PR TITLE
Check dependencies when enqueue via Queue.enqueue_job()

### LINKED ISSUE #1404: Job is executed before "depends_on" job is finished when using Job.create and q.enqueue_job
Consider the following snippet below and two running workers:
```python
r = Redis(
    host="localhost",
    port=6379,
    db=0
)

q = Queue(connection=r)

id_1 = str(uuid4())

q.enqueue(func_1, job_id=id_1)
q.enqueue(func_2, depends_on=id_1)
```
In that snippet, everything works as expected; both jobs are enqueued ( one on each worker ) but func_2 does not run before func_1 is finished. 

Now consider the following snippet, still with two workers:
```python
r = Redis(
    host="localhost",
    port=6379,
    db=0
)

q = Queue(connection=r)

id_1 = str(uuid4())
id_2 = str(uuid4())

job_1 = Job.create(func_1, id=id_1, connection=r) 
job_2 = Job.create(func_2, id=id_2, depends_on=id_1, connection=r)
q.enqueue_job(job_1)
q.enqueue_job(job_2)
```
When I run that code both jobs are started immediately, despite setting the `depends_on` argument. 

<br>
<br>

#### Mini bonus question
Why is it necessary to set `connection=` on the jobs when I use `Job.create`? Doesn't the `Queue` instance know what Redis connection to use already? Code fails without me setting `connection` on all the jobs as well.

### PR DESCRIPTION
**Pull request by sim6**
Fixes #1404

### CODE REVIEW

File: rq/queue.py [hunk 1]
```diff
@@ -663,11 +663,23 @@ def enqueue_call(
             on_success=on_success,
             on_failure=on_failure,
         )
+        return self._check_dependencies(job, pipeline=pipeline, at_front=at_front)
 
+    def _check_dependencies(self, job: 'Job', pipeline: Optional['Pipeline'] = None, at_front: bool = False) -> Job:
```
Comment (selwin) [inline rq/queue.py:668]: `_check_dependencies` shouldn't enqueue job to the queue. Besides, we already have `setup_dependencies`.

File: rq/queue.py [hunk 2]
```diff
@@ -663,11 +663,23 @@ def enqueue_call(
...
+            job (Job): The job to enqueue
+            pipeline (Optional[Pipeline], optional): The Redis pipeline to use. Defaults to None.
+            at_front (bool, optional): Whether should enqueue at the front of the queue. Defaults to False.
+
+        Returns:
+            Job: The enqued job
+        """
         job = self.setup_dependencies(job, pipeline=pipeline)
         # If we do not depend on an unfinished job, enqueue the job.
         if job.get_status(refresh=False) != JobStatus.DEFERRED:
```
Comment (selwin) [inline rq/queue.py:681]: I think we can simply move these checks to `enqueue_job` and that would solve the issue (`enqueue_job` should also setup dependencies), right?

File: rq/queue.py (hunk 1 above)
Comment (sim6) [reply rq/queue.py]: I changed the name from `_check_dependencies` to `_enqueue_checking_dependencies` to be more clear.

File: rq/queue.py (hunk 2 above)
Comment (sim6) [reply rq/queue.py:681]: Do you mean the `enqueue_job` added in this pull request or the original `enqueue_job`?

**Comment by sim6** [comment]
I improved new method name and rebased onto `master`.

**Comment by sim6** [comment]
Rebased to `master` to solve conflicts.

File: rq/queue.py (hunk 2 above)
Comment (selwin) [reply rq/queue.py:683]: I meant we should just add dependency checks to the original [enqueue_job()](https://github.com/rq/rq/blob/e92682c83aefda9475eba3dd9fe0ef42575d92b7/rq/queue.py#L982) method. So it looks something like this:
```python
def enqueue_job(self, job):
     job = self.setup_dependencies(job, pipeline=pipeline)
     # If we do not depend on an unfinished job, enqueue the job.
     if job.get_status(refresh=False) != JobStatus.DEFERRED:
        return self._enqueue_job(job, pipeline=pipeline, at_front=at_front)
    return job
```

Comment (sim6) [reply rq/queue.py:683]: I extracted the code from `enqueue_call()` to `_enqueue_checking_dependencies()` for use in `enqueue_call()` and `enqueue_job()`. Would you rather I duplicate the code?

**Comment by sim6** [comment]
Rebased onto `master`.

File: rq/queue.py (hunk 2 above)
Comment (selwin) [reply rq/queue.py:683]: Just move the logic in `_enqueue_checking_dependencies()` to `enqueue_job()`. `enqueue_call()` can then call `enqueue_job()`

Comment (sim6) [reply rq/queue.py:683]: I will do it. Thanks.

Comment (sim6) [reply rq/queue.py]: Done via https://github.com/rq/rq/compare/3f5e188fb93be91e9c7452d1a5724cdff9d5096a..4cf0db1ea9f2c93bbd2557882438efff3d2b77de

File: rq/queue.py (hunk 1 above)
Comment (sim6) [reply rq/queue.py]: Deleted via https://github.com/rq/rq/compare/3f5e188fb93be91e9c7452d1a5724cdff9d5096a..4cf0db1ea9f2c93bbd2557882438efff3d2b77de

**Comment by sim6** [comment]
Rebased onto `master`.

**Comment by sim6** [comment]
Test docstring fixed.

**Comment by selwin** [comment]
Thanks!

**Comment by sim6** [comment]
:)

**Pull request merged**